File: Day352/MaximumPorfitInJpobSchduling.go
package main

import (
	"fmt"
	"math"
	"sort"
)

type Pair struct {
	first, second int
}

type PairArray []Pair

func (a PairArray) Len() int {return len(a)}
func (a PairArray) Swap(i, j int) {a[i], a[j] = a[j], a[i]}
func (a PairArray) Less(i, j int) bool {return a[i].first < a[j].first}

func jobScheduling(startTime , endTime , profit []int) int {
	n := len(startTime)
	arr := make(PairArray, n)
	for i:=0; i<n; i++ {
		arr[i] = Pair{startTime[i], i}
	}
	sort.Sort(arr)
	dp := make([]int, n+1)
	for i:=n-1; i>=0; i-- {
		ni := sort.Search(n, func(j int) bool {
			return arr[j].first >= endTime[arr[i].second]
		})
		dp[i] = int(math.Max(float64(dp[i+1]), float64(profit[arr[i].second] + dp[ni])))
	}
	return dp[0]
}

func main() {
 	st1:= []int{1,2,3,3}
	et1:= []int{3,4,5,6}
	p1:= []int{50, 10, 40, 70}
	st2:= []int{1,2,3,4,6}
	et2:= []int{3,5,10,6, 9}
	p2:= []int{20, 20, 100, 70, 60}
	st3:= []int{1,1,1}
	et3:= []int{2,3,4}	
	p3:= []int{5,6,4}
	fmt.Println(jobScheduling(st1, et1, p1))
	fmt.Println(jobScheduling(st2, et2, p2))
	fmt.Println(jobScheduling(st3, et3, p3))
}

File: Day352/MaximumPorfitInJpobSchduling.java
package Day352;

import java.util.Arrays;
import java.util.Comparator;
import java.util.TreeMap;

// import javax.swing.RowFilter.ComparisonType;
// 
class Pair {
    int first, second;
    public Pair(int first, int second) {
        this.first = first;
        this.second = second;
    }
}

public class MaximumPorfitInJpobSchduling {
    public static int jobScheduling(int[] startTime, int[] endTime, int[] profit) {
        int n = startTime.length;
        int dp[][] = new int[n][3];
        for (int i=0; i<n; i++ ) {
            dp[i] = new int[]{startTime[i], endTime[i], profit[i]};
        }
        Arrays.sort(dp, (a,b)->a[1]-b[1]);
        TreeMap<Integer, Integer> map = new TreeMap<>();
        map.put(0, 0);
        for (int temp[]: dp) {
            int c = map.floorEntry(temp[0]).getValue() + temp[2];
            if (c > map.lastEntry().getValue()) {
                map.put(temp[1], c);
            }
        }
        return map.lastEntry().getValue();
    }
    public static void main(String[] args) {
        int st1[] = {1,2,3,3};
        int et1[] = {3,4,5,6};
        int p1[] = {50, 10, 40, 70};
        int st2[] = {1,2,3,4,6};
        int et2[] = {3,5,10,6, 9};
        int p2[] = {20, 20, 100, 70, 60};
        int st3[] = {1,1,1};
        int et3[] = {2,3,4};
        int p3[] = {5,6,4};
        System.out.println(jobScheduling(st1, et1, p1));
        System.out.println(jobScheduling(st2, et2, p2));
        System.out.println(jobScheduling(st3, et3, p3));
    }
}